### ISSUE TITLE
No duration for KubeNodeUnreachable alert

### ISSUE THREAD

**Issue by karanmagdani1** [kubernetes-monitoring/kubernetes-mixin#300]
The `KubeNodeUnreachable` alert does not have any duration for how long it should stay in pending. I see a lot of alerts which get resolved within seconds. 

https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/kubelet.libsonnet#L35

**Comment by brancz** [comment]
Could you share some graphs regarding this? I was kind of afraid that this would happen with this alert. It takes a bit though for the kubelet to be declared unreachable is my understanding so I also want to make sure we rule out something regarding your setup. Can you share some details about your cluster setup?

cc @smarterclayton

**Comment by smarterclayton** [comment]
What version of kube?  Do you have node leases on?  If not, what is your node refresh interval?  Is it all nodes?  Some nodes?  How much CPU pressure are those nodes under?

There is no normal scenario under which it is normal for nodes to be frequently unreachable, so if we can identify why you are seeing the alert that helps us identify whether you actually have a fairly serious issue (and whether we need to fix kube itsef)

**Comment by karanmagdani1** [comment]
Apologies for the late reply. 

- Kubernetes version: eks 1.14.6 
- The node lease is enabled by default and the lease duration is 40 seconds. 


This is not happening on all the nodes. Based off my investigations, this seems to be happening when a node is shut down and new node is starting up. 

These are some graphs for the node when the the alert triggers. 
![image](https://user-images.githubusercontent.com/46581504/70260431-1d4d2a00-1745-11ea-89ab-126877fd676b.png)

**Comment by karanmagdani1** [comment]
`Key=node.kubernetes.io/unreachable` is always preceded by `key=ToBeDeletedByClusterAutoscaler`
```
✔kube_node_spec_taint{effect="NoSchedule",endpoint="http",instance="172.24.xxx.xxx:8080",job="kube-state-metrics",key="ToBeDeletedByClusterAutoscaler",namespace="monitoring",node="ip-172-24-xxx-xx.ec2.internal",pod="prometheus-operator-kube-state-metrics-5d56987bnvbdz",service="prometheus-operator-kube-state-metrics",value="1576197116"}
```
```
✔kube_node_spec_taint{effect="NoSchedule",endpoint="http",instance="172.24.xxx.xx:8080",job="kube-state-metrics",key="node.kubernetes.io/unreachable",namespace="monitoring",node="ip-172-24-xxx-xx.ec2.internal",pod="prometheus-operator-kube-state-metrics-5d56987bnvbdz",service="prometheus-operator-kube-state-metrics"}
```

This leads me to believe that my cluster is running as expected. @smarterclayton

**Comment by richerve** [comment]
This is happening in our clusters too. The alerts started to trigger after the migrations to version 1.14.x, from 1.13.

In our case we use kops for provisioning and AWS spot instances to run jobs with very heavy memory/cpu usage. So those nodes come and go depending on the cronjobs scheduling time and resource consumption. Basically we have a high churn of nodes.

I wondering if it's related to a change in k8s 1.14+ were change of state is registered quicker or something like that.

**Comment by omerlh** [comment]
We also experience the same issue, I'll be happy to open a PR (adding the `for: 2m`) if this is the accepted solution.

**Comment by richerve** [comment]
Similar case here, around 2m is the max time that it takes for nodes to be fully terminated when using cluster-autoscaler. We decided to set that value to 5m as we have very big nodes and high churn, but setting at least a value of 2m by default would be great.

**Comment by amir-hadi** [comment]
@smarterclayton We are also using the cluster autoscaler and I believe the cluster is working properly. We have high churn on nodes as well, which means that this alert is happening quiet often. I'd also recommend to set a `for: 2m` at least, maybe even go with `for: 5m`. What do you think?

**Comment by smarterclayton** [comment]
Certainly any component that is terminating a node needs to unambiguously convey that the termination is normal, which should suppress the alert.  That would be more correct than extending the duration (extending the duration is a useful workaround but doesn’t solve the problem for clouds or envs that take longer than 2m).

I don’t believe there is a standardized “this is definitely going away taint”, but for instance even if there was we can’t suppress the alert because a node can crash during shutdown.

**Comment by smarterclayton** [comment]
I think the real bug here is that a node that is destined for termination is being marked unreachable.  That's a more involved fix, but setting `for` here is hacking around a fundamental gap in the node lifecycle (autoscaler is effectively defining its own arbitrary lifecycle).  

That fix is likely to look like autoscaler indicating via a standard signal (more standard than its current taint) that this machine is destined for shutdown and there is a grace period (5-10m) through which unreachable should be suppressed as an alert.

Until we have that, a 2m wait is acceptable (but I would push strongly against increasing it if asked by others).

**Comment by brancz** [comment]
Judging by the answer in https://github.com/kubernetes/autoscaler/issues/2811 I’m thinking we should probably already remove the for clause and make the alert ignore nodes tainted the way described there.

**Comment by richerve** [comment]
> Judging by the answer in [kubernetes/autoscaler#2811](https://github.com/kubernetes/autoscaler/issues/2811) I’m thinking we should probably already remove the for clause and make the alert ignore nodes tainted the way described there.

If I'm reading the comment in autoscaler correctly they say that although they set a custom taint, they can't control the behaviour when Kubernetes notices that the node is gone. So on this alert it wouldn't be the case that the node as the custom autoscaler's taint "ToBeDeletedByClusterAutoscaler" at some point in time but then k8s sets the Unreachable taint anyway? Wouldn't the alert trigger at that point if there's no minimal `for` condition?

**Comment by brancz** [comment]
Maybe I’m missing something. But why not only alert on nodes unreachable if they don’t have the taint? That should be doable in the alert expression.

**Comment by nvanheuverzwijn** [comment]
Would this be working ?
```
sum by (node) (kube_node_spec_taint{job="kube-state-metrics",key="node.kubernetes.io/unreachable",effect="NoSchedule"} unless kube_node_spec_taint{job="kube-state-metrics",key="ToBeDeletedByClusterAutoscaler",effect="NoSchedule"}) == 1
```

**Comment by chrisghill** [comment]
> Would this be working ?
> 
> ```
> sum by (node) (kube_node_spec_taint{job="kube-state-metrics",key="node.kubernetes.io/unreachable",effect="NoSchedule"} unless kube_node_spec_taint{job="kube-state-metrics",key="ToBeDeletedByClusterAutoscaler",effect="NoSchedule"}) == 1
> ```

We're running into this issue as well. I tried your expression, but in our case I don't think it works. Here is a snapshot of taint metrics right before a node was deleted by our cluster-autoscaler:
![image](https://user-images.githubusercontent.com/15616541/75584247-8d113f00-5a2c-11ea-9a2c-8c8cf78ca8f8.png)

Note that the `ToBeDeletedByClusterAutoscaler` metric and `node.kubernetes.io/unreachable` metric have a slightly different label set.  Specifically, `ToBeDeletedByClusterAutoscaler` adds a `value` metric (which I assume is a timestamp?) and the `unless` operator requires the metrics to have a matching label set.  I played around with the query a bit and I think this works:

```
(kube_node_spec_taint{job="kube-state-metrics",key="node.kubernetes.io/unreachable",effect="NoSchedule"} unless on(node) kube_node_spec_taint{job="kube-state-metrics",key="ToBeDeletedByClusterAutoscaler",effect="NoSchedule"}) == 1
```

Unfortunately I don't have a "true" positive to test against, but it does appear to filter out the `ToBeDeletedByClusterAutoscaler` false positive.
